### ISSUE TITLE
Add ability to extract latent representation from clustering algorithms

### ISSUE THREAD

**Issue by vemuribv** [WenjieDu/PyPOTS#177]
### 1. Feature description

The unsupervised clustering methods (VaDER & CRLI) should internally yield lower dimensional/latent representations of the input data that are used for the final clustering assignments. Users should be able to extract this latent representation for further downstream analysis.

### 2. Motivation

The ability to extract this latent representation would allow users to calculate [internal clustering validation measures](https://scikit-learn.org/stable/modules/clustering.html#silhouette-coefficient) like silhouette coefficient, gap statistic, and other indices. This is important especially in cases where there are no ground truth labels.

### 3. Your contribution

I am not yet totally clear on the clustering architecture implementations in PyPOTS (though I'm starting to familiarize myself more). However, I think these latent representations are already baked into the code:

- [Here](https://github.com/WenjieDu/PyPOTS/blob/4c08e051977a6591643c2f80a00694144d830f33/pypots/clustering/vader/model.py#L165) for VaDER
- [Here](https://github.com/WenjieDu/PyPOTS/blob/4c08e051977a6591643c2f80a00694144d830f33/pypots/clustering/crli/model.py#L438) for CRLI

It may be as simple as including these in what's returned after running .cluster (or provide another function to extract them solely).

**Comment by WenjieDu** [comment]
Hi there 👋,

Thank you so much for your attention to PyPOTS! You can follow me on GitHub to receive the latest news of PyPOTS. If you find PyPOTS helpful to your work, please star⭐️ this repository. Your star is your recognition, which can help more people notice PyPOTS and grow PyPOTS community. It matters and is definitely a kind of contribution to the community.

I have received your message and will respond ASAP. Thank you for your patience! 😃

Best,
Wenjie

**Comment by WenjieDu** [comment]
Hey @vemuribv, I'm going to adjust the PyPOTS framework API to make the clustering models return their latent representation. 

From your end, could you please give a thought to how PyPOTS can provide a more useful utility to help users calculate clustering validation measurements? e.g. could you help integrate some metrics you mentioned in [sklearn.clustering](https://scikit-learn.org/stable/modules/clustering.html#) into [pypots.utils.metrics](https://github.com/WenjieDu/PyPOTS/blob/main/pypots/utils/metrics.py)?

**Comment by vemuribv** [comment]
Sure thing, I'll work on that this week. Also, it would be great if the imputed array could be returned as well so we can do things like compare with the original array and generate cluster time series plots (see below from VaDER paper)

![image](https://github.com/WenjieDu/PyPOTS/assets/59896417/6e9b0bce-3a05-4695-b033-f0a087291273)

**Comment by WenjieDu** [comment]
No problem. I'll make the model to provide such an option to return the values. Could you please add the visualization functions as well?

**Comment by vemuribv** [comment]
Will do--working on something like this (PhysioNet (CRLI, k=3) here as an example):

![download](https://github.com/WenjieDu/PyPOTS/assets/59896417/275c140d-67fd-40b6-a3ac-a41027ee11e3)

![download](https://github.com/WenjieDu/PyPOTS/assets/59896417/3447a01b-9f5e-448b-bb3a-35ffe69bec50)

**Comment by WenjieDu** [comment]
Thanks for your PR #179, Bhargav! Will review it 😃

**Comment by WenjieDu** [comment]
Hey Bhargav, your PR #179 has been merged. Congrats! 👍

**Comment by vemuribv** [comment]
Awesome, thank you! What's the best place for the visualization functions? Also in pypots.utils.metrics?

**Comment by WenjieDu** [comment]
Absolutely my pleasure ;-) Please put them in `pypots/utils/visualization.py`, and `visualization.py` should be created by you.

**Comment by vemuribv** [comment]
Made the PR! Let me know if I can fix anything.

I also want to add visualizations like these below, but won't get to that until at least next week.

![image](https://github.com/WenjieDu/PyPOTS/assets/59896417/a821b8c9-1079-427b-a0e5-70484593e634)

![image](https://github.com/WenjieDu/PyPOTS/assets/59896417/ad9acf7f-4f9e-4d00-a49f-5d6912c65808)

**Comment by WenjieDu** [comment]
Great! I'm going to make the clustering models return their latent representation. Then you can write some unit tests for your functions.

**Comment by WenjieDu** [comment]
Hi Bhargav, I've made VaDER and CRLI return their latent representations for clustering as you requested in this issue. I've also written unit testing to test our functions of internal cluster validation metrics that you can refer to 

https://github.com/WenjieDu/PyPOTS/blob/09b494d7f26275a465339b903a83e031e6c32cef/tests/clustering/vader.py#L66-L76
